src/eos-shard-shard-file-impl-v1.c

```diff
@@ -0,0 +1,256 @@
+/* Copyright 2015 Endless Mobile, Inc. */
+
+/* This file is part of eos-shard.
+ *
+ * eos-shard is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * eos-shard is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-shard.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "eos-shard-shard-file-impl-v1.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "eos-shard-shard-file.h"
+#include "eos-shard-shard-file-impl.h"
+#include "eos-shard-enums.h"
+#include "eos-shard-blob.h"
+#include "eos-shard-record.h"
+#include "eos-shard-format-v1.h"
+
+struct _EosShardShardFileImplV1
+{
+  GObject parent;
+
+  EosShardShardFile *shard_file;
+  int fd;
+
+  GVariant *header_variant;
+};
+
+static void shard_file_impl_init (EosShardShardFileImplInterface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (EosShardShardFileImplV1, eos_shard_shard_file_impl_v1, G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (EOS_SHARD_TYPE_SHARD_FILE_IMPL, shard_file_impl_init))
+
+static void
+eos_shard_shard_file_impl_v1_finalize (GObject *object)
+{
+  EosShardShardFileImplV1 *self = EOS_SHARD_SHARD_FILE_IMPL_V1 (object);
+
+  g_clear_pointer (&self->header_variant, g_variant_unref);
+
+  G_OBJECT_CLASS (eos_shard_shard_file_impl_v1_parent_class)->finalize (object);
+}
+
+static void
+eos_shard_shard_file_impl_v1_class_init (EosShardShardFileImplV1Class *klass)
+{
+  GObjectClass *gobject_class = G_OBJECT_CLASS (klass);
+
+  gobject_class->finalize = eos_shard_shard_file_impl_v1_finalize;
+}
+
+static void
+eos_shard_shard_file_impl_v1_init (EosShardShardFileImplV1 *shard_file)
+{
+}
+
+/* XXX: Yes, I know this really should be using GInitable. */
```

Comment:
It still is -- but the impl isn't. There is technically wrong with this code, it's just a hygiene thing -- we shouldn't have constructors like this, for e.g. introspection reasons. But since it's an impl and an internal, it doesn't matter.